### ISSUE TITLE
Mapster.Async sometimes does not await all tasks

### ISSUE THREAD

**Issue by cmann-andagon** [MapsterMapper/Mapster#297]
I have a more complex mapping defined which uses multiple calls to `AfterMappingAsync`. Sometimes, `AdaptToTypeAsync()` does correctly await the complemation of all code defined in the different `AfterMappingAsync` calls.

I confirmed with the debugger that not all expected tasks are present in the task list in `AdaptToTypeAsync()`. E.g., I made 4 calls to `AfterMappingAsync()` but only two tasks are registered in the task list.

Note, that I am also using the dependency injection plugin and I use it to access a DataManager (connection to database) which is provided by the dependency injection.

**Comment by chaowlert** [comment]
Could you provide working sample? All tasks seemed to be tracked. Make sure you `await`  all async methods in `AfterMappingAsync`.

**Comment by cmann-andagon** [comment]
From inside the AfterMappingAsync(), I access MongoDB. Both Mapster and the MongoDB driver use of AsyncLocal. My guess is, that this causes some conflict.

I found this article https://nelsonparente.medium.com/a-little-riddle-with-asynclocal-1fd11322067f, which states that there might be issues if you try to update the data inside an AsyncLocal.

Unfortunately, I do not have time to create a working sample right now and it might be a bit tricky to get the timing right.

**Comment by cmann-andagon** [comment]
I was finally able to reproduce the issue reliably. The issue occurs when `AdaptToTypeAsync()` is used inside `AfterMappingAsync()`. I have created a pull request with a test which reproduces the issue: https://github.com/MapsterMapper/Mapster/pull/310.

It looks like the issue is located in the `MapContextScope`. As it does not allocate a new `MapContext` everytime, it can happen that the special parameter which holds the tasks is overwritten and some tasks which should be awaited are not properly registered. I have attempted a fix in https://github.com/MapsterMapper/Mapster/pull/311, but this breaks 4 other tests.
